Implement new output variable Surface Internal User Specified Location Temperature for Construction:InternalSource

RKStrand opened the pull request:
## Pull request overview

For radiant systems that use a Construction:InternalSource, EnergyPlus has a feature that allows the user to report a temperature internal to the construction.  This is useful for a variety of reasons and also for non-radiant systems potentially.  The problem was that though the construction definition made it possible for the user to request this temperature the code to use this information properly in the code was lacking.  This defect fix corrects this problem and now EnergyPlus properly produces output that shows the internal temperature.  This issue was logged as GitHub Issue #3478.
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [ x ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [ x ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs GitHub Issue #3478
### Review Checklist

This will not be exhaustively relevant to every PR.
- [x] Code style (parentheses padding, variable names)
- [x] Functional code review (it has to work!)
- [x] If defect, results of running current develop vs this branch should exhibit the fix
- [x] CI status: all green or justified
- [x] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [x] Argument types
  - [x] If any virtual classes, ensure virtual destructor included, other things
- n/a IDD changes:
- [x] If new idf included, locally check the err file and other outputs
- [x] Documentation changes in place
- [x] Changed docs build successfully

RKStrand committed:
Fix of defect

The user specified location temperature was not being calculated for
low temperature radiant systems.  This fix implemented the calculation
of this value.  Testing included asking for the temperature at the
source location and comparing it to the temperature already calculated
at that location. This got the exact same result for all time steps.
When the location of the user specified temperature was changed, a
different value between the source location and the zone air
temperature was obtained.  This value was noted to vary in how close it
was to the other values based on whether or not the radiant system was
on.  Both of these tests were successful.

RKStrand committed:
Merge branch 'develop' into 113835831-Source-Temp-Not-Reporting

RKStrand committed:
Unit test, doc addition, additional code fixes

This commit includes the unit test to make sure that the temperature at
the user location is being calculated properly and shifted during the
update process.  It also includes a minor docs addition for the output
variable and two related fixes.  This commit completes the work.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 113835831-Source-Temp-Not-Reporting

mjwitte committed:
Internal source temperature - correct variable name in SetUpOutput

mjwitte commented:
@Myoldmopar Should we be listing new output variables in the output rules doc?

Myoldmopar commented:
No, I don't think that's necessary.

mjwitte commented:
Tested this, fixed the variable name in the SetupOutputVariable call, output values look reasonable. Will merge after CI comes back clean.

eringold commented:
@mjwitte @RKStrand Just FYI, the Construction:InternalSource section in the I/O reference is missing all output descriptions and some of the inputs (instead there are a bunch of HAMT construction outputs). I noted this in the big documentation issue here: https://github.com/NREL/EnergyPlus/issues/5603#issuecomment-226952335, but it may have flown under the radar.
If it wasn't and a fix is in the works, please disregard!

mjwitte commented:
@eringold Thanks for pointing that out.  The new output variable here was simply added to the existing list in the general surface output section.  Will attempt to clean this up here.

mjwitte committed:
I/O Ref cleanup HAMT and Construction:Internal source outputs

mjwitte commented:
@eringold The HAMT and Construction:InternalSource output sections have been cleaned up (duplicates gone, internal source outputs moved from prior location, outdated material deleted).  Please review [here](http://energyplus.s3-website-us-east-1.amazonaws.com/assets/2016-08-31-EnergyPlus-ee02cc0bef-x86_64-gcc-4.8/) and let me know if it looks ok.

mjwitte commented:
@RKStrand This is impacting [performance](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-ee02cc0befbab1dc6207e8dda177184f9d5750b3-x86_64-Linux-Ubuntu-14.04-gcc-4.8.html#performance) with a 0.4% increase in run time.  Small, but not insignificant.  I see some places where operations related to source temperature/flux/histories are inside an if construct.SourceSinkPresent but not everywhere.  Seems we could avoid these arrays (new and old source variables) entirely for runs without any sources by adding a few ifs and may be a new global AnySourceSinkPresent flag.

eringold commented:
@mjwitte looks great, thank you!

RKStrand commented:
@mjwitte I'll look at it and see if I can construct some if-thens to improve things.  I don't want the performance hit unless we absolutely cannot avoid it.  As you mentioned, 0.4% is not huge but to high for such an issue.  If you need to switch this back to unfinished in Pivotal Tracker and reassign to me, that is fine.

RKStrand committed:
Changes to address speed issue

The implementation of this defect fix ended up resulting in a 0.4%
speed increase in the test suite.  Many changes were made to try to
eliminate this speed issue while still maintaining all of the
functionality.  Now, variables won’t be calculated, initialized, or
even allocated if they won’t be used.

RKStrand committed:
Corrections for unit test issue

One of the unit tests was crashing and the fix for this involved
updating the clear_state and changing one of the if-thens to resolve
that.

mjwitte commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -3978,13 +4018,17 @@ namespace HeatBalanceSurfaceManager {
-				OpaqSurfInsFaceConduction( SurfNum ) = surface.Area * QH_12;
-				OpaqSurfInsFaceConductionFlux( SurfNum ) = QH_12; //CR 8901
+				if ( construct.SourceSinkPresent ) {
+					Real64 const QH_12s = QH[ l21 ] = QH_12 + QsrcHist( SurfNum, 1 ) * construct.CTFSourceIn( 0 );
```
@RKStrand I'm guessing that the small diffs for most files with internal source are coming from here.  Not a problem, really, just a place where the sequence of calculations changed slightly.

mjwitte commented:
@RKStrand Nice!  Now we have a 0.28% _positive_ speed improvement.

RKStrand commented:
Whoo-hoo!  Do I get extra credit for such a speed improvement?

> On Sep 2, 2016, at 2:39 PM, Michael J. Witte notifications@github.com wrote:
> 
> @RKStrand https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_RKStrand&d=CwMCaQ&c=8hUWFZcy2Z-Za5rBPlktOQ&r=yqH1mHrRNln4fYsdbjkCiwPoHQDib3PyYUYzBDdv_IQ&m=yZe6nsvQ4EdoI6Nhzi7Wxg-fUP2Y5IogC8SOJ3EXzPk&s=qrMoaDN2g38Vf55ylwt_to04gyR0Xlg9TMDy5zADobc&e= Nice! Now we have a 0.28% positive speed improvement.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_NREL_EnergyPlus_pull_5810-23issuecomment-2D244469214&d=CwMCaQ&c=8hUWFZcy2Z-Za5rBPlktOQ&r=yqH1mHrRNln4fYsdbjkCiwPoHQDib3PyYUYzBDdv_IQ&m=yZe6nsvQ4EdoI6Nhzi7Wxg-fUP2Y5IogC8SOJ3EXzPk&s=EehDxwW6Hbynrdayp3Uvz929p_pWBl7SOBScxd4JSmY&e=, or mute the thread https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AFiTvZ92XKfI71ZFUGnVp1QeGsFtwvXSks5qmHtUgaJpZM4JhsAv&d=CwMCaQ&c=8hUWFZcy2Z-Za5rBPlktOQ&r=yqH1mHrRNln4fYsdbjkCiwPoHQDib3PyYUYzBDdv_IQ&m=yZe6nsvQ4EdoI6Nhzi7Wxg-fUP2Y5IogC8SOJ3EXzPk&s=6wavWo74B2Z5gtxq_Z6612GQAUYCJvnmLiwjjGmzqbc&e=.

RKStrand commented:
@mjwitte It looks like the pull request is still active.  I have marked Pivotal Tracker back to finished/delivered and switched the owner back to you.  Am I correct in assuming this pull request is still active and that you will finish the review?

mjwitte commented:
Yes, sir.
